src/v/cloud_roles/azure_vm_refresh_impl.cc

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/azure_vm_refresh_impl.h"
+
+#include "cloud_roles/request_response_helpers.h"
+#include "json/ostreamwrapper.h"
+#include "json/schema.h"
+
+#include <rapidjson/error/en.h>
+
+namespace cloud_roles {
+
+azure_vm_refresh_impl::azure_vm_refresh_impl(
+  net::unresolved_address address,
+  aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    std::move(address), std::move(region), as, retry_params) {}
+
+auto azure_vm_refresh_impl::print(std::ostream& os) const -> std::ostream& {
+    fmt::print(os, "azure_vm_refresh_impl{{address:{}}}", address());
+    return os;
+}
+
+auto azure_vm_refresh_impl::fetch_credentials() -> ss::future<api_response> {
+    auto client_id_opt = config::shard_local_cfg()
+                           .cloud_storage_azure_managed_identity_id.value();
+    if (unlikely(!client_id_opt.has_value())) {
+        // This implementation requires client_id from config to be set.
+        // Strictly speaking, IMDS service does not require it if there is only
+        // a system-assigned managed identity or only one user-assigned managed
+        // identity but it would be brittle to not specify it, it could break if
+        // the user added a new one. This verification is also performed in
+        // admin/server.cc::patch_cluster_config
+        vlog(
+          clrl_log.error,
+          "missing cloud_storage_azure_managed_identity_id in config");
```

Comment:
does it make sense to use a default / default-generated client id if none of this works without the client id configuration being set?

Reply:
cool that makes sense.

i'm trying to reconcile just one thing. you mention above regarding using a default some scenarios related to getting an attached identity from a pod. presumably that would be a unique id per pod? the part i'm trying to resolve is this: if it was important to mention pods because unique id is important, then will we have that problem with the implementation in this PR because it uses a single id per cluster taken from the configuration?

Reply:
thanks for the explanation!